File: results/2026-04-07_225702/tasks/16-environment-matrix-generator/default-opus/generated-code/matrix_generator.py
"""
Environment matrix generator for GitHub Actions strategy.matrix.

Generates a build matrix (JSON) from a configuration of dimensions (OS options,
language versions, feature flags), with support for include/exclude rules,
max-parallel limits, fail-fast, and matrix size validation.
"""

from __future__ import annotations

import json
import sys
import itertools
from dataclasses import dataclass, field
from typing import Any


class MatrixError(Exception):
    """Raised when matrix configuration is invalid or exceeds constraints."""


@dataclass
class MatrixConfig:
    """Configuration for generating a GitHub Actions build matrix."""
    dimensions: dict[str, list[Any]] = field(default_factory=dict)
    include: list[dict[str, Any]] = field(default_factory=list)
    exclude: list[dict[str, Any]] = field(default_factory=list)
    fail_fast: bool | None = None
    max_parallel: int | None = None
    max_combinations: int = 256  # GitHub Actions default limit


def _count_cartesian_size(dimensions: dict[str, list[Any]]) -> int:
    """Count the total cartesian product size from dimensions."""
    if not dimensions:
        return 0
    size = 1
    for values in dimensions.values():
        size *= len(values)
    return size


def _count_excluded(dimensions: dict[str, list[Any]], excludes: list[dict[str, Any]]) -> int:
    """Count how many combos from the cartesian product each exclude rule removes.

    An exclude rule matches all combos where the specified keys have the specified
    values. For keys not mentioned in the exclude, all values match.
    """
    if not dimensions or not excludes:
        return 0

    dim_keys = list(dimensions.keys())
    total_excluded = 0

    for exc in excludes:
        # For each exclude rule, figure out how many combos it removes.
        # For each dimension key: if the key is in the exclude, it matches 1 value;
        # otherwise it matches all values in that dimension.
        matched = 1
        for key in dim_keys:
            if key in exc:
                # Only removes combos where this key equals the excluded value
                if exc[key] in dimensions[key]:
                    matched *= 1
                else:
                    # The excluded value isn't even in the dimension, so 0 matches
                    matched = 0
                    break
            else:
                matched *= len(dimensions[key])
        total_excluded += matched

    return total_excluded


def _validate_config(config: MatrixConfig) -> None:
    """Run all validations on the config before generating the matrix."""
    if not config.dimensions and not config.include:
        raise MatrixError("At least one dimension or include entry is required")

    for dim_name, dim_values in config.dimensions.items():
        if not dim_values:
            raise MatrixError(f"Dimension '{dim_name}' must have at least one value")
        if len(dim_values) != len(set(str(v) for v in dim_values)):
            raise MatrixError(f"Dimension '{dim_name}' has duplicate values")

    if config.max_parallel is not None and config.max_parallel < 1:
        raise MatrixError("max_parallel must be a positive integer")


def _validate_matrix_size(config: MatrixConfig) -> None:
    """Validate the effective matrix size doesn't exceed the maximum."""
    base_size = _count_cartesian_size(config.dimensions)
    excluded = _count_excluded(config.dimensions, config.exclude)
    # Include entries that introduce entirely new combos (not augmenting existing)
    extra_includes = len(config.include)
    effective = base_size - excluded + extra_includes

    if effective > config.max_combinations:
        raise MatrixError(
            f"Matrix has {effective} combinations exceeds maximum of "
            f"{config.max_combinations}"
        )


def generate_matrix(config: MatrixConfig) -> dict[str, Any]:
    """Generate a GitHub Actions strategy.matrix object from config."""
    _validate_config(config)

    # Build the matrix section
    matrix: dict[str, Any] = {}
    for dim_name, dim_values in config.dimensions.items():
        matrix[dim_name] = dim_values

    # Add include entries — these add extra combos or augment existing ones
    if config.include:
        matrix["include"] = config.include

    # Add exclude entries — these remove matching combos from the product
    if config.exclude:
        matrix["exclude"] = config.exclude

    # Validate matrix size doesn't exceed the limit
    _validate_matrix_size(config)

    result: dict[str, Any] = {"matrix": matrix}

    # Strategy-level options (siblings of matrix, not inside it)
    if config.fail_fast is not None:
        result["fail-fast"] = config.fail_fast
    if config.max_parallel is not None:
        result["max-parallel"] = config.max_parallel

    return result


def matrix_to_json(config: MatrixConfig, indent: int = 2) -> str:
    """Generate matrix and return as formatted JSON string."""
    return json.dumps(generate_matrix(config), indent=indent)


def load_config_from_dict(data: dict[str, Any]) -> MatrixConfig:
    """Create a MatrixConfig from a plain dictionary (e.g. parsed from JSON/YAML)."""
    return MatrixConfig(
        dimensions=data.get("dimensions", {}),
        include=data.get("include", []),
        exclude=data.get("exclude", []),
        fail_fast=data.get("fail_fast"),
        max_parallel=data.get("max_parallel"),
        max_combinations=data.get("max_combinations", 256),
    )


def _matches_exclude(combo: dict[str, Any], exclude: dict[str, Any]) -> bool:
    """Check if a combination matches an exclude rule (all specified keys must match)."""
    return all(combo.get(k) == v for k, v in exclude.items())


def expand_combinations(config: MatrixConfig) -> list[dict[str, Any]]:
    """Enumerate all concrete job combinations after applying include/exclude rules.

    This is useful for debugging or inspecting what the matrix will actually run.
    """
    combos: list[dict[str, Any]] = []

    # Generate cartesian product of all dimensions
    if config.dimensions:
        keys = list(config.dimensions.keys())
        for values in itertools.product(*config.dimensions.values()):
            combos.append(dict(zip(keys, values)))

    # Remove excluded combinations
    if config.exclude:
        combos = [
            c for c in combos
            if not any(_matches_exclude(c, exc) for exc in config.exclude)
        ]

    # Add include entries as extra combinations
    combos.extend(config.include)

    return combos


def main() -> None:
    """CLI entry point: reads a JSON config from stdin or a file argument,
    generates the matrix, and prints the complete JSON to stdout."""
    if len(sys.argv) > 1:
        filepath = sys.argv[1]
        try:
            with open(filepath) as f:
                raw = json.load(f)
        except FileNotFoundError:
            print(f"Error: file not found: {filepath}", file=sys.stderr)
            sys.exit(1)
        except json.JSONDecodeError as e:
            print(f"Error: invalid JSON in {filepath}: {e}", file=sys.stderr)
            sys.exit(1)
    else:
        try:
            raw = json.load(sys.stdin)
        except json.JSONDecodeError as e:
            print(f"Error: invalid JSON from stdin: {e}", file=sys.stderr)
            sys.exit(1)

    try:
        config = load_config_from_dict(raw)
        print(matrix_to_json(config))
    except MatrixError as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: results/2026-04-17_004319/tasks/16-environment-matrix-generator/typescript-bun-sonnet-medium/generated-code/matrix-generator.ts
// Environment Matrix Generator
// Generates GitHub Actions strategy.matrix JSON from a configuration file.

export interface MatrixConfig {
  dimensions: Record<string, string[]>;
  include?: Record<string, string>[];
  exclude?: Record<string, string>[];
  maxParallel?: number;
  failFast?: boolean;
  maxSize?: number;
}

// The matrix property mirrors what GitHub Actions expects in strategy.matrix
export interface GHAMatrix {
  [key: string]: string[] | Record<string, string>[] | undefined;
  include?: Record<string, string>[];
  exclude?: Record<string, string>[];
}

export interface MatrixOutput {
  matrix: GHAMatrix;
  maxParallel?: number;
  failFast?: boolean;
  combinations: number;
  valid: boolean;
  errors?: string[];
}

const DEFAULT_MAX_SIZE = 256;

// Compute the Cartesian product of all dimensions.
// Returns an array of every combination as a flat object.
export function cartesianProduct(
  dims: Record<string, string[]>
): Record<string, string>[] {
  const keys = Object.keys(dims);
  if (keys.length === 0) return [{}];

  let result: Record<string, string>[] = [{}];

  for (const key of keys) {
    const values = dims[key];
    const next: Record<string, string>[] = [];
    for (const existing of result) {
      for (const value of values) {
        next.push({ ...existing, [key]: value });
      }
    }
    result = next;
  }

  return result;
}

// Check whether a combination matches all key/value pairs in a rule.
function matchesRule(
  combo: Record<string, string>,
  rule: Record<string, string>
): boolean {
  return Object.entries(rule).every(([k, v]) => combo[k] === v);
}

// Determine how many total jobs the matrix will run:
// - Start with Cartesian product
// - Remove excluded combos
// - For each include: if it matches an existing combo (on dimension keys only),
//   it's an augmentation (no new job); otherwise it's a new job.
function countEffectiveCombinations(
  baseCombos: Record<string, string>[],
  includes: Record<string, string>[],
  dimensionKeys: string[]
): number {
  let count = baseCombos.length;

  for (const include of includes) {
    // Extract only the dimension keys from this include entry
    const dimPart = Object.fromEntries(
      dimensionKeys
        .filter((k) => include[k] !== undefined)
        .map((k) => [k, include[k]])
    );

    // If dimPart is non-empty, check if it matches any existing combo
    const isAugmentation =
      Object.keys(dimPart).length > 0 &&
      baseCombos.some((combo) => matchesRule(combo, dimPart));

    if (!isAugmentation) {
      count++;
    }
  }

  return count;
}

export function generateMatrix(config: MatrixConfig): MatrixOutput {
  const { dimensions, include, exclude, maxParallel, failFast, maxSize } =
    config;

  const limit = maxSize ?? DEFAULT_MAX_SIZE;
  const dimensionKeys = Object.keys(dimensions);

  // Build the base matrix from the Cartesian product
  let baseCombos =
    dimensionKeys.length === 0 ? [] : cartesianProduct(dimensions);

  // Apply exclude rules
  if (exclude && exclude.length > 0) {
    baseCombos = baseCombos.filter(
      (combo) => !exclude.some((rule) => matchesRule(combo, rule))
    );
  }

  // Count total effective combinations (base + any new jobs from includes)
  const totalCombinations = countEffectiveCombinations(
    baseCombos,
    include ?? [],
    dimensionKeys
  );

  // Build the GHA matrix object
  const matrix: GHAMatrix = { ...dimensions };
  if (include && include.length > 0) {
    matrix.include = include;
  }
  if (exclude && exclude.length > 0) {
    matrix.exclude = exclude;
  }

  // Validate against max size
  const errors: string[] = [];
  if (totalCombinations > limit) {
    errors.push(
      `Matrix has ${totalCombinations} combinations, exceeding max-size of ${limit}`
    );
  }

  const output: MatrixOutput = {
    matrix,
    combinations: totalCombinations,
    valid: errors.length === 0,
  };

  if (maxParallel !== undefined) output.maxParallel = maxParallel;
  if (failFast !== undefined) output.failFast = failFast;
  if (errors.length > 0) output.errors = errors;

  return output;
}

// CLI entry point: bun run matrix-generator.ts <config-file.json>
async function main() {
  const args = process.argv.slice(2);
  if (args.length === 0) {
    console.error("Usage: bun run matrix-generator.ts <config-file.json>");
    process.exit(1);
  }

  const configPath = args[0];
  let config: MatrixConfig;

  try {
    const raw = await Bun.file(configPath).text();
    config = JSON.parse(raw) as MatrixConfig;
  } catch (err) {
    console.error(`Failed to read config file '${configPath}': ${err}`);
    process.exit(1);
  }

  const result = generateMatrix(config);
  console.log(JSON.stringify(result, null, 2));

  if (!result.valid) {
    console.error(`Matrix validation failed: ${result.errors?.join(", ")}`);
    process.exit(1);
  }
}

// Run CLI when invoked directly (not imported as a module)
if (import.meta.main) {
  main();
}
